_drafts/best-and-worst-gcc-clang-compiler-flags.md

```diff
@@ -0,0 +1,892 @@
+EXTRA_CFLAGS="-Wall -Werror -Wno-error=memset-elt-size" make
+main.c: In function 'prv_various_memset_bugs':
+main.c:47:3: error: 'memset' used with constant zero length parameter; this could be due to transposed parameters [-Werror=memset-transposed-args]
+   memset(buf, NUM_ITEMS, 0x0);
+   ^~~~~~
+main.c:48:3: warning: 'memset' used with length equal to number of elements without multiplication by element size [-Wmemset-elt-size]
+memset(buf, 0x0, NUM_ITEMS);
+
+# Don't enable any warnings for memset-elt-size
+EXTRA_CFLAGS="-Wall -Werror -Wno-memset-elt-size" make
+main.c: In function 'prv_various_memset_bugs':
+main.c:47:3: error: 'memset' used with constant zero length parameter; this could be due to transposed parameters [-Werror=memset-transposed-args]
+   memset(buf, NUM_ITEMS, 0x0);
+   ^~~~~~
+```
+
+### -Weverything (Clang only)
+
+By design, Clang generally matches the set of compiler flag options available in the GNU toolchain, but there are a few that are different.
+
+For embedded projects, it's useful to cross-compile the source code with Clang to surface additional
+issues that GCC may have missed.
+
+One flag there are some _strong opinions_[^2] about is `-Weverything` which literally enables every
+single warning available to Clang! Many suggest you should _never_ flip on this flag in a
+build. The strategy I prefer, followed by other projects such as CppUTest[^3], is to enable the
+warning but liberally disable warnings from it that are not useful.
+
+I've found this to be a pretty neat way to discover new warning flags built into clang, such as
+`-Wdocumentation`, which can catch errors in comments! For example,
+
+```c
+//! A bad docstring that clang can identify issues with using -Wdocumentation
+//!
+//! @param a The first value of the summation
+//! @param c The second value of the summation
+//!
+//! @return the sum of a & b
+int simple_math_get_sum(int a, int b);
+```
+
+```bash
+USE_CLANG=1 EXTRA_CFLAGS=-Weverything make
+[...]
+./simple_math.h:6:12: warning: parameter 'c' not found in the function declaration [-Wdocumentation]
+//! @param c The second value of the summation
+           ^
+./simple_math.h:6:12: note: did you mean 'b'?
+//! @param c The second value of the summation
+           ^
+           b
+```
+
+## The Best One-Off Warning Options (options not covered by -Wall or -Wextra)
+
+There's many more warnings and checks that are _not_ enabled by `-Wall` or `-Wextra`. In this section we
+will explore some of the best ones.
+
+### -Wshadow
+
+Shadowing variables at the very least makes code difficult to read and often can be indicative of a
+bug because the code is not operating on the value the programmer expects. These can easily be
+detected by using `-Wshadow`.
+
+```c
+int variable_shadow_error_example2(void) {
+  int result = 4;
+
+  for (int i = 0; i < 10; i++) {
+    int result = i;
+    // do something with result
+  }
+
+  return result;
+}
+```
+
+```bash
+EXTRA_CFLAGS=-Wshadow make
```

Comment:
When you show these shell commands, prefix every one of them with `$ `. The commands don't stick out super well.